### ISSUE TITLE
Indirectly changing values within switch cases causes an 'unintentional comparison' error

### ISSUE THREAD

**Issue by ephf** [microsoft/TypeScript#60326]
### 🔎 Search Terms

switch, case, indirect, unintentional, comparison, change, equality, update, modify, if

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I could not find FAQ entries related to the issue.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.7.0-beta#code/DYUwLgBAHhC8EFYDcAoAZgVwHYGMwEsB7LCHACwEMsBzEAfSgAoBKCAb2jggEYAmJCAF8UKAM4B3fGHKMorNighLSFUSEQAuRcp3kqtBi1TadS-Gllx4feRAD0AKggA6VxAd2hKQUA

### 💻 Code

```ts
let x = 5;
function change_x() { x = 12; }

switch(x) {
    case 5:
        change_x();

        if(x == 12) { /* ... */ }
}
```

### 🙁 Actual behavior

```
bug.ts:8:12 - error TS2367: This comparison appears to be unintentional because the types '5' and '12' have no overlap.

8         if(x == 12) { /* ... */ }
             ~~~~~~~
```

This error should not show when compiling as `x` is set to `12` when calling the `change_x` function.

### 🙂 Expected behavior

It is expected that there is no error given when compiling.

### Additional information about the issue

This can be navigated around by introducing a second variable before the switch statement.

```ts
let x = 5;
function change_x() { x = 12; }

let y = x; // here
switch(y) {
    case 5:
        change_x();

        if(x == 12) { /* ... */ }
}
```

This code no longer produces an error in the latest version

*see [playground](https://www.typescriptlang.org/play/?ts=5.7.0-beta#code/DYUwLgBAHhC8EFYDcAoAZgVwHYGMwEsB7LCHACwEMsBzEAfSgAoBKCAb2jggEYAmJCAF8UKUJACeXKKgDOAd3xhyjcazYoIm0hRkhEALg1bj5KrQYtUR45vxpGMWPD5qIAegBUEAHS+IHtyEUQSA)*

**Comment by MartinJohns** [comment]
Duplicate of #9998.
